media-proxy/include/mesh/st2110tx.h

```diff
@@ -0,0 +1,168 @@
+            return -1;
+        }
+
+        inet_pton(AF_INET, cfg_st2110.remote_ip_addr, ops.port.dip_addr[MTL_PORT_P]);
+        ops.port.udp_port[MTL_PORT_P] = cfg_st2110.remote_port;
+        strlcpy(ops.port.port[MTL_PORT_P], dev_port.c_str(), MTL_PORT_MAX_LEN);
+        ops.port.udp_src_port[MTL_PORT_P] = cfg_st2110.local_port;
+        ops.port.num_port = 1;
+
+        char session_name[NAME_MAX] = "";
+        snprintf(session_name, NAME_MAX, "mcm_mtl_tx_%d", session_id);
+        if (ops.name)
+            free((void *)ops.name);
+        ops.name = strdup(session_name);
+        ops.framebuff_cnt = 4;
+
+        ops.priv = this; // app handle register to lib
+        ops.notify_frame_available = frame_available_cb;
+
+        log::info("ST2110Tx: configure")("port", ops.port.port[MTL_PORT_P])(
+            "dip_addr", std::to_string(ops.port.dip_addr[MTL_PORT_P][0]) + " " +
+                            std::to_string(ops.port.dip_addr[MTL_PORT_P][1]) + " " +
+                            std::to_string(ops.port.dip_addr[MTL_PORT_P][2]) + " " +
+                            std::to_string(ops.port.dip_addr[MTL_PORT_P][3]))(
+            "num_port", ops.port.num_port)("udp_port", ops.port.udp_port[MTL_PORT_P])(
+            "udp_src_port", ops.port.udp_src_port[MTL_PORT_P])("name", ops.name)("framebuff_cnt",
+                                                                                 ops.framebuff_cnt);
+
+        return 0;
+    }
+
+    Result on_establish(context::Context& ctx) override {
+        _ctx = context::WithCancel(ctx);
+        stop = false;
+
+        mtl_session = create_session(mtl_device, &ops);
+        if (!mtl_session) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    };
+
+    Result on_shutdown(context::Context& ctx) override {
+        _ctx.cancel();
+
+        if (mtl_session) {
+            close_session(mtl_session);
+            mtl_session = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    Result on_receive(context::Context& ctx, void *ptr, uint32_t sz, uint32_t& sent) override {
+        int to_be_sent = std::min(transfer_size, sz);
+        // TODO: add error/warning if sent is different than _transfer_size
+
+        FRAME *frame = NULL;
+        do {
+            // Get empty buffer from MTL
+            frame = get_frame(mtl_session);
+            if (!frame) {
+                std::unique_lock lk(mx);
+                cv.wait(lk, _ctx.stop_token(), [this] { return stop.load(); });
+                stop = false;
+                if (_ctx.cancelled()) {
+                    return set_result(Result::error_shutdown);
+                }
+            }
+        } while (!frame);
+
+        if (frame) {
```

Comment:
it will be always true as the while above exits only when frame != NULL

Reply:
Fixed